{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# biopython\n",
    "- [url](https://github.com/biopython/biopython)\n",
    "- FASTQ\n",
    "- FASTA\n",
    "- SAM\n",
    "- BAM\n",
    "- VCF\n",
    "- BED"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def _opener(filename):\n",
    "    if filename.endswith('.gz'):\n",
    "        import gzip\n",
    "        return gzip.open\n",
    "    elif filename.endswith('.bz2'):\n",
    "        import bz2\n",
    "        return bz2.open\n",
    "    elif filename.endswith('.lzma'):\n",
    "        import lzma\n",
    "        return lzma.open\n",
    "    else:\n",
    "        return open\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "size=218\n",
      "key=ATTC,v=3\n",
      "key=TGCG,v=3\n",
      "key=TGGC,v=7\n",
      "key=TGCC,v=7\n",
      "key=CCTG,v=19\n",
      "key=TCCA,v=9\n",
      "key=CCTT,v=10\n",
      "key=GGTG,v=14\n",
      "key=CAGT,v=8\n",
      "key=ACAC,v=4\n",
      "key=TGAC,v=2\n",
      "key=GTTT,v=5\n",
      "key=CTGA,v=3\n",
      "key=TCTT,v=6\n",
      "key=ACAA,v=6\n",
      "key=CCCA,v=23\n",
      "key=TGTA,v=7\n",
      "key=CTGC,v=5\n",
      "key=TCCG,v=3\n",
      "key=CCAA,v=12\n",
      "key=ACTA,v=1\n",
      "key=TCAA,v=8\n",
      "key=CCCT,v=21\n",
      "key=CCCG,v=4\n",
      "key=TGGA,v=7\n",
      "key=CTAC,v=3\n",
      "key=TAAG,v=5\n",
      "key=CATT,v=3\n",
      "key=GTCT,v=2\n",
      "key=TTGT,v=3\n",
      "key=ACCT,v=6\n",
      "key=ACAG,v=5\n",
      "key=GCAC,v=6\n",
      "key=CACT,v=5\n",
      "key=GCAT,v=4\n",
      "key=CAAG,v=9\n",
      "key=GGCT,v=9\n",
      "key=TACC,v=6\n",
      "key=CACA,v=8\n",
      "key=ACTT,v=5\n",
      "key=TGAG,v=10\n",
      "key=ACTG,v=5\n",
      "key=CCAG,v=28\n",
      "key=CAGA,v=6\n",
      "key=CCAC,v=11\n",
      "key=CTTG,v=5\n",
      "key=ATTT,v=3\n",
      "key=TGTT,v=12\n",
      "key=CCGT,v=2\n",
      "key=TTCA,v=3\n",
      "key=GGTT,v=6\n",
      "key=ATCT,v=1\n",
      "key=CGGC,v=5\n",
      "key=CTTT,v=11\n",
      "key=TCTG,v=7\n",
      "key=AGAA,v=10\n",
      "key=TACT,v=3\n",
      "key=TAGG,v=8\n",
      "key=GCCT,v=10\n",
      "key=AATT,v=4\n",
      "key=AGCA,v=3\n",
      "key=TGTC,v=7\n",
      "key=TCTA,v=2\n",
      "key=TCAT,v=7\n",
      "key=GCAA,v=5\n",
      "key=CAAA,v=4\n",
      "key=AACA,v=3\n",
      "key=GCAG,v=8\n",
      "key=CCTC,v=20\n",
      "key=CTAT,v=7\n",
      "key=TGAT,v=12\n",
      "key=CGAG,v=1\n",
      "key=GGAA,v=10\n",
      "key=ACAT,v=4\n",
      "key=CAGC,v=5\n",
      "key=AGGT,v=2\n",
      "key=GGGG,v=6\n",
      "key=TTTC,v=4\n",
      "key=CCTA,v=9\n",
      "key=TGCA,v=5\n",
      "key=TAAA,v=9\n",
      "key=AAGA,v=4\n",
      "key=TTAA,v=2\n",
      "key=GAGC,v=4\n",
      "key=TAGC,v=3\n",
      "key=TATT,v=5\n",
      "key=CGCT,v=2\n",
      "key=TTTG,v=5\n",
      "key=TATG,v=7\n",
      "key=GCTG,v=11\n",
      "key=GGAC,v=4\n",
      "key=GAAA,v=6\n",
      "key=CTTA,v=4\n",
      "key=GCCA,v=12\n",
      "key=GTAT,v=1\n",
      "key=GGTC,v=9\n",
      "key=TTCT,v=3\n",
      "key=GCTT,v=7\n",
      "key=CTGG,v=3\n",
      "key=GATA,v=3\n",
      "key=GCCC,v=8\n",
      "key=TTCG,v=1\n",
      "key=CATA,v=3\n",
      "key=ATAG,v=2\n",
      "key=TGTG,v=9\n",
      "key=TTTT,v=9\n",
      "key=AAGG,v=3\n",
      "key=TCCC,v=3\n",
      "key=AAGC,v=1\n",
      "key=AAGT,v=1\n",
      "key=AAAC,v=3\n",
      "key=CAAT,v=6\n",
      "key=GATC,v=4\n",
      "key=TACA,v=7\n",
      "key=CATG,v=10\n",
      "key=CGGG,v=2\n",
      "key=GTGA,v=1\n",
      "key=CGTG,v=1\n",
      "key=AGGA,v=4\n",
      "key=CCAT,v=10\n",
      "key=CGAC,v=5\n",
      "key=GTCC,v=2\n",
      "key=CACG,v=3\n",
      "key=GCCG,v=3\n",
      "key=AGTG,v=1\n",
      "key=GCGC,v=1\n",
      "key=CATC,v=5\n",
      "key=TGAA,v=8\n",
      "key=ATCC,v=1\n",
      "key=CACC,v=3\n",
      "key=TGGT,v=3\n",
      "key=TTAC,v=2\n",
      "key=TGGG,v=7\n",
      "key=CCCC,v=20\n",
      "key=GGCA,v=7\n",
      "key=CTAG,v=2\n",
      "key=TCAC,v=3\n",
      "key=TAAC,v=2\n",
      "key=TCAG,v=4\n",
      "key=AAAG,v=4\n",
      "key=ATAA,v=1\n",
      "key=CTCT,v=5\n",
      "key=ATAC,v=2\n",
      "key=CAGG,v=5\n",
      "key=CTGT,v=1\n",
      "key=AGCT,v=2\n",
      "key=GATT,v=3\n",
      "key=CGGT,v=2\n",
      "key=ATTG,v=1\n",
      "key=GAAG,v=7\n",
      "key=AAAT,v=3\n",
      "key=CTAA,v=3\n",
      "key=GGAT,v=4\n",
      "key=GGCC,v=2\n",
      "key=GTTA,v=1\n",
      "key=AGAG,v=3\n",
      "key=GGAG,v=7\n",
      "key=GCTC,v=4\n",
      "key=AGGC,v=2\n",
      "key=CAAC,v=4\n",
      "key=GAGG,v=1\n",
      "key=GGTA,v=2\n",
      "key=TATC,v=3\n",
      "key=GTAG,v=1\n",
      "key=ATGA,v=1\n",
      "key=GGGA,v=4\n",
      "key=CGAT,v=1\n",
      "key=GCTA,v=4\n",
      "key=GTGT,v=1\n",
      "key=CCGA,v=2\n",
      "key=CGCC,v=1\n",
      "key=ACCA,v=4\n",
      "key=TTGC,v=2\n",
      "key=ACTC,v=4\n",
      "key=AGTC,v=1\n",
      "key=CTTC,v=2\n",
      "key=ATAT,v=2\n",
      "key=TCCT,v=1\n",
      "key=GTTC,v=1\n",
      "key=ATGC,v=1\n",
      "key=GGGT,v=2\n",
      "key=TATA,v=5\n",
      "key=AGTA,v=1\n",
      "key=TAGT,v=1\n",
      "key=TAAT,v=4\n",
      "key=GAGT,v=1\n",
      "key=AGTT,v=3\n",
      "key=AGCC,v=1\n",
      "key=CGTT,v=3\n",
      "key=TTCC,v=2\n",
      "key=AACT,v=3\n",
      "key=TGCT,v=4\n",
      "key=GACC,v=2\n",
      "key=GTAA,v=1\n",
      "key=ATGT,v=2\n",
      "key=AGAT,v=1\n",
      "key=TTGA,v=3\n",
      "key=GTTG,v=5\n",
      "key=AGGG,v=2\n",
      "key=GACG,v=2\n",
      "key=GAAT,v=1\n",
      "key=TCTC,v=4\n",
      "key=TCGA,v=1\n",
      "key=GAGA,v=2\n",
      "key=GTCA,v=1\n",
      "key=CGTC,v=1\n",
      "key=CTCC,v=1\n",
      "key=ACCC,v=1\n",
      "key=GATG,v=1\n",
      "key=ACCG,v=1\n",
      "key=GACA,v=1\n",
      "key=AAAA,v=1\n",
      "key=TTTA,v=1\n",
      "key=CGAA,v=1\n",
      "key=TAGA,v=2\n",
      "key=CTCA,v=1\n",
      "key=ATCA,v=1\n",
      "key=CCGC,v=1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/tom/local/anaconda3/lib/python3.6/site-packages/Bio/Seq.py:152: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "# FASTQ by Biopython\n",
    "\n",
    "from Bio.SeqIO.QualityIO import FastqPhredIterator, FastqPhredWriter\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "def head(fpath, h = 10) :\n",
    "    openfun = _opener(fpath)\n",
    "    if openfun is not None :\n",
    "        with openfun(fpath) as fh:\n",
    "            try :\n",
    "                it = FastqPhredIterator(fh)\n",
    "                for i, r in enumerate(it) :\n",
    "                    if i > 10:\n",
    "                        break\n",
    "                    else:\n",
    "                        print(\"i={}, r={}\".format(i,r))\n",
    "                              \n",
    "            except Exception as e:\n",
    "                    print(\"Error : {}\".format(str(e)))\n",
    "                    \n",
    "                    \n",
    "                              \n",
    "def gen(fpath, num = 100):\n",
    "    openfun = _opener(fpath)\n",
    "    if openfun is not None :\n",
    "        with openfun(fpath) as fh:\n",
    "            it = FastqPhredIterator(fh)\n",
    "            for i, r in enumerate(it) :\n",
    "                if i < num :\n",
    "                    yield  r\n",
    "                else :\n",
    "                    break\n",
    "        \n",
    "            \n",
    "def sample(src, frac = 0.1) :\n",
    "    import random\n",
    "    percent = int(frac * 100)\n",
    "    for r in src:\n",
    "        if random.randrange(1,101) < percent:\n",
    "            yield r\n",
    "\"\"\"\n",
    "    - \"qual\" means simple quality files using PHRED scores (e.g. from Roche 454)\n",
    "    - \"fastq\" means Sanger style FASTQ files using PHRED scores and an ASCII\n",
    "      offset of 33 (e.g. from the NCBI Short Read Archive and Illumina 1.8+).\n",
    "      These can potentially hold PHRED scores from 0 to 93.\n",
    "    - \"fastq-sanger\" is an alias for \"fastq\".\n",
    "    - \"fastq-solexa\" means old Solexa (and also very early Illumina) style FASTQ\n",
    "      files, using Solexa scores with an ASCII offset 64. These can hold Solexa\n",
    "      scores from -5 to 62.\n",
    "    - \"fastq-illumina\" means newer Illumina 1.3 to 1.7 style FASTQ files, using\n",
    "      PHRED scores but with an ASCII offset 64, allowing PHRED scores from 0\n",
    "      to 62.\n",
    "      \n",
    ">>> from Bio import SeqIO\n",
    ">>> for record in SeqIO.parse(\"Quality/example.fastq\", \"fastq\"):\n",
    "...     print(\"%s %s\" % (record.id, record.seq))\n",
    "\n",
    "\"\"\"\n",
    "def output(fpath, src) :\n",
    "    with _opener(fpath)(fpath, \"w\") as out :\n",
    "        writer = FastqPhredWriter(out)\n",
    "        writer.write_header()\n",
    "    \n",
    "        for r in src:\n",
    "            #print(r)\n",
    "            writer.write_record(r)\n",
    "            \n",
    "def stat(it) :\n",
    "    import collections\n",
    "    \n",
    "    index = collections.defaultdict(int)\n",
    "    for r in it:\n",
    "        index[r.seq[0:4]]+=1\n",
    "    print(\"size={}\".format(len(index)))\n",
    "    for k in index.keys() :\n",
    "        print(\"key={},v={}\".format(k,index[k]))\n",
    "        \n",
    "    \n",
    "src = gen(\"SRR3503015.fastq\", 1000)\n",
    "stat(src)\n",
    "#output(\"srr3503015_sample.fastq\", src)\n",
    "\n",
    "#head(\"SRR3503015.fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: SRR3503015.1\n",
      "Name: SRR3503015.1\n",
      "Description: SRR3503015.1 1 length=194\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('ATTCATTCAAGAAATGTTAAACATGCAAATTTTTAGCACATGTATATGTAATAT...ATT', SingleLetterAlphabet())\n",
      "[25, 35, 35, 37, 39, 39, 39, 39, 39, 41, 40, 40, 41, 40, 39, 41, 37, 34, 40, 39, 39, 40, 41, 34, 37, 39, 39, 40, 40, 40, 39, 40, 36, 38, 38, 37, 39, 39, 40, 40, 38, 40, 38, 35, 39, 40, 41, 40, 37, 37, 38, 38, 40, 29, 37, 39, 38, 39, 39, 36, 39, 39, 40, 37, 37, 39, 40, 37, 39, 37, 37, 34, 34, 39, 39, 40, 40, 35, 36, 39, 38, 40, 40, 41, 39, 39, 38, 37, 34, 36, 37, 33, 35, 36, 37, 37, 37, 32, 35, 35, 35, 39, 39, 39, 39, 39, 40, 37, 39, 39, 39, 38, 40, 38, 40, 39, 39, 37, 39, 40, 38, 39, 40, 41, 39, 39, 39, 38, 40, 41, 41, 40, 38, 35, 34, 40, 37, 39, 40, 40, 36, 39, 40, 35, 37, 39, 40, 36, 39, 38, 39, 40, 40, 40, 40, 35, 31, 37, 39, 38, 38, 39, 40, 38, 38, 38, 40, 40, 40, 41, 41, 41, 41, 34, 38, 39, 40, 39, 36, 34, 39, 39, 38, 39, 37, 37, 37, 37, 35, 31, 36, 34, 36, 36]\n",
      "phred_quality\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "f = SeqIO.parse(\"SRR3503015.fastq\", \"fastq\")\n",
    "r = next(f)\n",
    "print(r)\n",
    "ano = r.letter_annotations\n",
    "print(r.letter_annotations[\"phred_quality\"])\n",
    "for i in ano:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'biopython'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-9d707d9d190f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mbiopython\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'biopython'"
     ]
    }
   ],
   "source": [
    "import biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR3503015.1 1 length=194\r\n",
      "ATTCATTCAAGAAATGTTAAACATGCAAATTTTTAGCACATGTATATGTAATATTCACTTTACTACATATTGCAATGTACTTGAAGCTTTTATATTAGAAGACTTAATAATGAATGCTAATACCTTTCCTATATGAAGGATTTTTCATGATAGACATTGAAATATGAAATTTTGATGGTACTATTATTCCCATT\r\n",
      "+\r\n",
      ":DDFHHHHHJIIJIHJFCIHHIJCFHHIIIHIEGGFHHIIGIGDHIJIFFGGI>FHGHHEHHIFFHIFHFFCCHHIIDEHGIIJHHGFCEFBDEFFFADDDHHHHHIFHHHGIGIHHFHIGHIJHHHGIJJIGDCIFHIIEHIDFHIEHGHIIIID@FHGGHIGGGIIIJJJJCGHIHECHHGHFFFFD@ECEE\r\n",
      "@SRR3503015.2 2 length=194\r\n",
      "TGCGCCCTGCAGCGGCAGCAGGCGACGGCAGCCGAGAAGCCCTGGGATGCCAAGGCCCCCGAGGGCCGGCTGCGGAAGTGTGAGAGCACCGACTCGGCCCGGTCTTCCGGGGCCGCACGTTGTCCAGCTGGGCATCGTCCACCACCGCCTGGTTGTGGGAGATGAGCTGGGCGATGCGCTCCTCCAGCCGCTTC\r\n",
      "+\r\n",
      "DDEDHHHHHIJJJJJIJJJIHIGGGJHGEDDCC>B?BD<CBDCBDDDDDDDDDBDDDBBB;775BDB9BB><@<B<@B:A:ACD@C<AAA<<><@8?BDFFHHHHHJJHIJJJJIJJJJHIIJJJJHJJHHHHFEFFCCDDDDBBDBDD;BD:AD7?BDBD;ACCDDCDDDBBB<>BDB>BDCCD<ACDDD599\r\n",
      "@SRR3503015.3 3 length=194\r\n",
      "TGGCAGTCATAGTATGGTTCTTCTACCTGAATCTCTTGATCTTCACCTTCTTCTGGGTCTTCAATTCCCACACCGTCAGGCTCGTCGGCATCTCCCTGGCCGAGCAGCCAAATGGGGATGCTGATGAGCCAGATGGTGCAGGTATTGAAGACCCAGAAGAAAGAGCTGAAGAGCCAGAGGGAAAAGCTGAAGAG\r\n",
      "+\r\n",
      "ADDDFADFFEGFA:A<@<E>CFEHF@CAHFFFGEFDFDEFGFIEF9D9BFD?BBDFI9==C@FEFFFFIEFCEFFEBDDD;6@?;<28?9594>@?(DDDDDDHGDFGGGIJ>FGIIBHGIIEEBHGGHIGIDBGDBBC<C=./=F@CC=D@EECA=;CFFECCCEDCDCDCDCDDBDDDD0?B?CBCC>@::2\r\n",
      "@SRR3503015.4 4 length=194\r\n",
      "TGCCAGAAGGTCTGTCACTGTCACCGTGGTGTTGGTGAGTCCAAACTGTCGAGGGAGGAAGCGGACATTTGGGCTGCATGGCTCACACTGTTTTATAAAGAAAGGATAGTTATTTACCTCATTCAGTTCCATTGTAAATTATCTATTTCCCTTGTAGTAGAAAACAGAAGTGAGGCTCATTAATCTTTGTTGAA\r\n",
      "+\r\n",
      "DDFFGHHHHJFHHIIJJJJJJJIJJJGHIEGGGIJFHGHIIIJGFIJJJJJHIJJGHHBBFFDD?@BCDEECBDDDCDDDDDDDCDCDDDDDDDCD4DFFFHHHHHIJJIJJJJJJJJJJJJJJJJJJJJIJJJHGIJIJIJJGGIIJIIIIJJIJIJGJIJJJIJIJIJ@GHJJJJHHHHHHFEFFFFEECCC\r\n",
      "@SRR3503015.5 5 length=194\r\n",
      "CCTGGCTGCCAAACTCGTCCTCGCGGGAGACATGGTTGGCCCGCCTAGGTGGCTCCTCATCCTCCTCACAGTCATAGTCATCCAGGATGGGAGTCTCTGCCGCTGGGTCCCCAGGCGCGGGAGAGCTTCCGTTCATCCCGGCTGCAAACCCACAATTCCGTCATTGGCGTGCCCATCCGGGAGACTCCCATCCT\r\n",
      "+\r\n",
      "DFFFHHHHHJJJHHJJJJJJJJJJJJJIIJJJJJJHIGHFFFDDDDDDD@CDDDDDDDDDDDDDDDDDDDDDCCEEDEEEEDDDDDDDDDDDD?CDDBDF>FHHHHJHJJJJJJJJJJJJJGIEHJJHHHHFFFFFFEDDDDD=@CDDDDDDDDDDDDDDDDDDEDDDDDD<BD@CCB<BB<@9<CCDD<@CD3\r\n",
      "@SRR3503015.6 6 length=194\r\n",
      "TCCAAAGGAAGTGAGGGGTCCTTCTATGGTGAAACTAATTTGTGATTCATGTGTTCCCTCATCTGAGAATCTGGGGCAGGCCCATTTCCATGACCCTGGCTGACCTGCGCAACGTGATCAAGACACTGTTGAAAGCTGGCCACATAGCATGGACTTCTTCTGGAAAAGGGTCATGGAAATGGGCCTGCCCCAGA\r\n",
      "+\r\n",
      "DFFFHHHGHJJHIJJJJJIJJIJIIIJIIFHIJIJJJJJJJIIIHIJJJJJJJJJJJJJJJJJJJJIJJHHCHFHFFDDDDDDBDDDDEEEDCCDDDADDFHHHHHJJJJJIJJ:CHIIJJJJJJJJJJJGIIJIJJJJJJJJIHJIJJJJHHHHHHFFFFFFDEEEEDD?CDDDDCDDDDDDDDDDDDDDBD8\r\n",
      "@SRR3503015.7 7 length=194\r\n",
      "CCTTTTAAACCAGAGTCGAACTTTACTGTCTCCTCAGCAACAACAGCAGCAGCAAGTGGCACTTGGCCCTGGCATGCCAGCAAAGCCTCTTCAACACGGCTGGGTCTACGGTGTTTTGTTCCTTGCCCGTCAGGAGGAGGGTTGGACCCAGGGCTCCAGGGCTAGAAAAGTGTTGAAGAGGCTTTGCTGGCATG\r\n",
      "+\r\n",
      "DDFFHHGHGJJJJJJHIJJJJJJJIIJJGJIJJGIHIJIJIJJIGGGIJJJJIIJI@GHIIJIHHHHHFFDBCEEEEDDDDDDDCCDDDDDDEDDDBADFFHHHFHIIJJJFGHIJJJIJIJJJIJJJIIJIIJFHIFHHI;E@ECEHFFFDDCEDDDDDDDBDDCDCDD>AACBACDCDB@?CDCCCAC@CC<\r\n",
      "@SRR3503015.8 8 length=194\r\n",
      "GGTGGGGACAGGGGCGATGCTTCAGGTGGTGGGGATAGAGGCGTCTCAAGTGCAGGAGATGGGGGTGACTCTTCCGGTGGAGACAAGGGCGACTCCTCACCTGAGGAGTCACCCCTGTCCCCACCACCTGAGGAATCACCCACGTCCCCACCACCTGAGGCATCACGCCTGTCACCACCACCTGAGGAATTGCC\r\n",
      "+\r\n",
      "DDBDHHHHHJJJJJIJJJJJJJJIJJ8BH@GIHFEDEECEEDDBDDDEDDEDDDDDDDBDDDDDD9BDDDDCCDDDDDDDDDDDDDDDDDDDDDDDCDDFFHHHHHJJJJJJJJJJJJJJJJJJJJJJJJJJJJIJIJJJJJJHIJJJJJHHHFFFFEEDDBDDDDDDDDDDDDDDDDDDDDDDACDDDDDDDD\r\n",
      "@SRR3503015.9 9 length=194\r\n",
      "CAGTACATGCACTTCAGCACCTTCTGGGCGTCTTCCTTCCCTTCCATTTCCAGCAAGGAGCGTTTGCGAGGCTTGGACCAGCGCTTGGGGTTGTTGTCGCTGTAAGGACTGCAGCGCTGCCTACGACACCCTGGTGGAGTTGACGGTGCACATGAACGAGACGGGGCATTACCGCGACGACAACCATGAGACCG\r\n",
      "+\r\n",
      "DDFFHHHHHJIJJJFIJJJJJJJGIJJIGHHGHIGGIIJJIJJGHFHHIHGIIJJJJJEHIGEFFFFDBDDBBCB?CDDDDDDBDDBDDD59@>BDADDFFHHGHHJJJIJIJJJJIIJIJJIJJIJJJJJJJJCGIGIGHHHCDFBBDEEDDDCDDDDDDDDDDDBBCCDDDBDD9BDBDBDDDDCCCACC@>\r\n",
      "@SRR3503015.10 10 length=194\r\n",
      "ACACTGACCCATATGCTCAGCCCCCATTGACTCCTCGGCCCCAACCTCCGCCCCCTGAGAGCTGCTGTGCTCTGCCCCCTCGCTCACTGCCCTCCGCTAAAGGCAACTTCAGGGGGCTGGGGTCGGGGTGGCTTATGCAATGGGGCAAATGGGTCACGGGACTGAGGGCGTGAGGGTGGGCGAGAATAAGGGTC\r\n",
      "+\r\n",
      "ADEFHFHHGJJJJJJIJJIIJIJJJHHIGHJEHGDHGIIHIJCHJIIGIAGIGFFAB>ACDB;>AC@@ACA:@:ACBDD50&9@@7CC>CCD@BD9&FFFFHHHHHJJIJGIJJJJJIJJJJDGIJJI;=E;BFDEEEDECED?BB@DDDDDD>CBD>BD;>ABCCBDDD>B?BDD2?BD?B9@99@CCCCD<>\r\n"
     ]
    }
   ],
   "source": [
    "! cat srr3503015_sample.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
