{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# biopython\n",
    "- [biopython](https://github.com/biopython/biopython)\n",
    "- FASTQ\n",
    "- FASTA\n",
    "- SAM\n",
    "- BAM\n",
    "- VCF\n",
    "- BED"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FASTQ"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nfor i, s in enumerate(src) :\\n    if i < 1 :\\n        print(s)\\n        print(s[6:-6])\\n        print(s.seq)\\n        print(s[6:-6].seq)\\n'"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# FASTQ by Biopython\n",
    "\n",
    "from Bio.SeqIO.QualityIO import FastqPhredIterator, FastqPhredWriter\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "\n",
    "def _opener(filename):\n",
    "    if filename.endswith('.gz'):\n",
    "        import gzip\n",
    "        return gzip.open\n",
    "    elif filename.endswith('.bz2'):\n",
    "        import bz2\n",
    "        return bz2.open\n",
    "    elif filename.endswith('.lzma'):\n",
    "        import lzma\n",
    "        return lzma.open\n",
    "    else:\n",
    "        return open\n",
    "    \n",
    "\n",
    "def head(fpath, h = 10) :\n",
    "    openfun = _opener(fpath)\n",
    "    if openfun is not None :\n",
    "        with openfun(fpath) as fh:\n",
    "            try :\n",
    "                it = FastqPhredIterator(fh)\n",
    "                for i, r in enumerate(it) :\n",
    "                    if i > 10:\n",
    "                        break\n",
    "                    else:\n",
    "                        print(\"i={}, r={}\".format(i,r))\n",
    "                              \n",
    "            except Exception as e:\n",
    "                    print(\"Error : {}\".format(str(e)))\n",
    "                    \n",
    "                    \n",
    "                              \n",
    "def gen(fpath, num = 0):\n",
    "    openfun = _opener(fpath)\n",
    "    if openfun is not None :\n",
    "        with openfun(fpath) as fh:\n",
    "            it = FastqPhredIterator(fh)\n",
    "            for i, r in enumerate(it) :\n",
    "                if i < num or num < 1:\n",
    "                    yield  r\n",
    "                else :\n",
    "                    break\n",
    "                    \n",
    "\n",
    "def clip(src, head = 6, tail = 6):\n",
    "    for r in src :\n",
    "        yield r[head :- tail]\n",
    "        \n",
    "            \n",
    "def sample(src, frac = 0.1) :\n",
    "    import random\n",
    "    percent = int(frac * 100)\n",
    "    for r in src:\n",
    "        if random.randrange(1,101) < percent:\n",
    "            yield r\n",
    "\"\"\"\n",
    "    - \"qual\" means simple quality files using PHRED scores (e.g. from Roche 454)\n",
    "    - \"fastq\" means Sanger style FASTQ files using PHRED scores and an ASCII\n",
    "      offset of 33 (e.g. from the NCBI Short Read Archive and Illumina 1.8+).\n",
    "      These can potentially hold PHRED scores from 0 to 93.\n",
    "    - \"fastq-sanger\" is an alias for \"fastq\".\n",
    "    - \"fastq-solexa\" means old Solexa (and also very early Illumina) style FASTQ\n",
    "      files, using Solexa scores with an ASCII offset 64. These can hold Solexa\n",
    "      scores from -5 to 62.\n",
    "    - \"fastq-illumina\" means newer Illumina 1.3 to 1.7 style FASTQ files, using\n",
    "      PHRED scores but with an ASCII offset 64, allowing PHRED scores from 0\n",
    "      to 62.\n",
    "      \n",
    ">>> from Bio import SeqIO\n",
    ">>> for record in SeqIO.parse(\"Quality/example.fastq\", \"fastq\"):\n",
    "...     print(\"%s %s\" % (record.id, record.seq))\n",
    "\n",
    "\"\"\"\n",
    "def output(fpath, src) :\n",
    "    with _opener(fpath)(fpath, \"w\") as out :\n",
    "        writer = FastqPhredWriter(out)\n",
    "        writer.write_header()\n",
    "    \n",
    "        for r in src:\n",
    "            #print(r)\n",
    "            writer.write_record(r)\n",
    "            \n",
    "            \n",
    "def split(read1, read2, src) :\n",
    "    with _opener(read1)(read1, \"w\") as f1:\n",
    "        writer1 = FastqPhredWriter(f1)\n",
    "        writer1.write_header()\n",
    "        \n",
    "        with _opener(read2)(read2, \"w\") as f2 :\n",
    "            writer2 = FastqPhredWriter(f2)\n",
    "            writer2.write_header()\n",
    "            \n",
    "            for r in src :\n",
    "                l = int(len(r.seq)/2)\n",
    "                #print(l)\n",
    "                writer1.write_record(r[0:l])\n",
    "                writer2.write_record(r[l:])\n",
    "                \n",
    "                \n",
    "def stat(it) :\n",
    "    import collections\n",
    "    \n",
    "    index = collections.defaultdict(int)\n",
    "    for r in it:\n",
    "        index[r.seq[0:4]]+=1\n",
    "        index[r.seq[-4:]]+=1\n",
    "    print(\"size={}\".format(len(index)))\n",
    "    for k in index.keys() :\n",
    "        print(\"key={},v={}\".format(k,index[k]))\n",
    "        \n",
    "    \n",
    "'''\n",
    "SRR3503011\tSAMN04478409\tCTR15\tCTR15-c1-noPhiX\n",
    "SRR3503012\tSAMN04478409\tCTR15\tCTR15-c1\n",
    "SRR3503013\tSAMN04478410\tCTR31\tCTR31-c1-NextSeq\n",
    "SRR3503014\tSAMN04478410\tCTR31\tCTR31-c1-noPhiX\n",
    "SRR3503015\tSAMN04478410\tCTR31\tCTR31-c1\n",
    "'''\n",
    "\n",
    "num = 10000\n",
    "src = gen(\"SRR3503015.fastq\", num)\n",
    "split(\"SRR15-{}_1.fastq\".format(num), \"SRR15-{}_2.fastq\".format(num), src)\n",
    "\n",
    "\"\"\"\n",
    "for i, s in enumerate(src) :\n",
    "    if i < 1 :\n",
    "        print(s)\n",
    "        print(s[6:-6])\n",
    "        print(s.seq)\n",
    "        print(s[6:-6].seq)\n",
    "\"\"\"\n",
    "        \n",
    "#stat(clip(src))\n",
    "#output(\"srr3503015_sample.fastq\", clip(src))\n",
    "\n",
    "#head(\"SRR3503015.fastq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "range=chr13:48916201-48917522=4\n",
      "range=chr17:7576020-7577221=4\n"
     ]
    }
   ],
   "source": [
    "from Bio.SeqIO.QualityIO import FastqPhredIterator, FastqPhredWriter\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "from collections import defaultdict\n",
    "\n",
    "ids = defaultdict(int)\n",
    "with open(\"capp-seq/selector-clean.fa\", \"w\") as handle:\n",
    "    for r in SeqIO.parse(\"capp-seq/selector-orig.fa\", \"fasta\"):\n",
    "        desc = r.description.split(\" \")\n",
    "        r.id = desc[1]\n",
    "        r.name = desc[1]\n",
    "        ids[r.id] += 1\n",
    "        if ids[r.id] >1 :\n",
    "            continue\n",
    "        #print(r.id)\n",
    "        SeqIO.write(r, handle, \"fasta\")\n",
    "        \n",
    "for key in ids.keys() :\n",
    "    if ids[key] > 1 :\n",
    "        print(\"{}={}\".format(key,ids[key]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# BED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['/home/tom/zdata/bioref/chromFa-hg19/chr1.fa', '/home/tom/zdata/bioref/chromFa-hg19/chr2.fa']\n",
      "{} len = {} chr1 249250621\n",
      "{} len = {} chr2 243199373\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "hg19_root = os.environ[\"BIOREF\"] + \"/chromFa-hg19\"\n",
    "#names = [str(i) for i in range(1,23)] + [\"X\" , \"Y\" , \"M\"]\n",
    "names = [str(i) for i in range(1,3)]\n",
    "chrs = [\"chr\" + i for i in names]\n",
    "paths = { i: hg19_root + '/' + i + \".fa\" for i in chrs}\n",
    "print(list(paths.values()))\n",
    "index_db = SeqIO.index_db(hg19_root+\"/index_db.sqlite\", list(paths.values()), \"fasta\")\n",
    "#index = { i: SeqIO.index(paths[i], \"fasta\") for i in chrs}\n",
    "#print(chrs)\n",
    "#for i in chrs :\n",
    "#    print( paths[i])\n",
    "for k in index_db.keys() :\n",
    "    print(\"{} len = {}\", k , len(index_db[k].seq))#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR3503015.1 1 length=194\r\n",
      "ATTCATTCAAGAAATGTTAAACATGCAAATTTTTAGCACATGTATATGTAATATTCACTTTACTACATATTGCAATGTACTTGAAGCTTTTATATTAGAAGACTTAATAATGAATGCTAATACCTTTCCTATATGAAGGATTTTTCATGATAGACATTGAAATATGAAATTTTGATGGTACTATTATTCCCATT\r\n",
      "+\r\n",
      ":DDFHHHHHJIIJIHJFCIHHIJCFHHIIIHIEGGFHHIIGIGDHIJIFFGGI>FHGHHEHHIFFHIFHFFCCHHIIDEHGIIJHHGFCEFBDEFFFADDDHHHHHIFHHHGIGIHHFHIGHIJHHHGIJJIGDCIFHIIEHIDFHIEHGHIIIID@FHGGHIGGGIIIJJJJCGHIHECHHGHFFFFD@ECEE\r\n",
      "@SRR3503015.2 2 length=194\r\n",
      "TGCGCCCTGCAGCGGCAGCAGGCGACGGCAGCCGAGAAGCCCTGGGATGCCAAGGCCCCCGAGGGCCGGCTGCGGAAGTGTGAGAGCACCGACTCGGCCCGGTCTTCCGGGGCCGCACGTTGTCCAGCTGGGCATCGTCCACCACCGCCTGGTTGTGGGAGATGAGCTGGGCGATGCGCTCCTCCAGCCGCTTC\r\n",
      "+\r\n",
      "DDEDHHHHHIJJJJJIJJJIHIGGGJHGEDDCC>B?BD<CBDCBDDDDDDDDDBDDDBBB;775BDB9BB><@<B<@B:A:ACD@C<AAA<<><@8?BDFFHHHHHJJHIJJJJIJJJJHIIJJJJHJJHHHHFEFFCCDDDDBBDBDD;BD:AD7?BDBD;ACCDDCDDDBBB<>BDB>BDCCD<ACDDD599\r\n",
      "@SRR3503015.3 3 length=194\r\n",
      "TGGCAGTCATAGTATGGTTCTTCTACCTGAATCTCTTGATCTTCACCTTCTTCTGGGTCTTCAATTCCCACACCGTCAGGCTCGTCGGCATCTCCCTGGCCGAGCAGCCAAATGGGGATGCTGATGAGCCAGATGGTGCAGGTATTGAAGACCCAGAAGAAAGAGCTGAAGAGCCAGAGGGAAAAGCTGAAGAG\r\n",
      "+\r\n",
      "ADDDFADFFEGFA:A<@<E>CFEHF@CAHFFFGEFDFDEFGFIEF9D9BFD?BBDFI9==C@FEFFFFIEFCEFFEBDDD;6@?;<28?9594>@?(DDDDDDHGDFGGGIJ>FGIIBHGIIEEBHGGHIGIDBGDBBC<C=./=F@CC=D@EECA=;CFFECCCEDCDCDCDCDDBDDDD0?B?CBCC>@::2\r\n",
      "@SRR3503015.4 4 length=194\r\n",
      "TGCCAGAAGGTCTGTCACTGTCACCGTGGTGTTGGTGAGTCCAAACTGTCGAGGGAGGAAGCGGACATTTGGGCTGCATGGCTCACACTGTTTTATAAAGAAAGGATAGTTATTTACCTCATTCAGTTCCATTGTAAATTATCTATTTCCCTTGTAGTAGAAAACAGAAGTGAGGCTCATTAATCTTTGTTGAA\r\n",
      "+\r\n",
      "DDFFGHHHHJFHHIIJJJJJJJIJJJGHIEGGGIJFHGHIIIJGFIJJJJJHIJJGHHBBFFDD?@BCDEECBDDDCDDDDDDDCDCDDDDDDDCD4DFFFHHHHHIJJIJJJJJJJJJJJJJJJJJJJJIJJJHGIJIJIJJGGIIJIIIIJJIJIJGJIJJJIJIJIJ@GHJJJJHHHHHHFEFFFFEECCC\r\n",
      "@SRR3503015.5 5 length=194\r\n",
      "CCTGGCTGCCAAACTCGTCCTCGCGGGAGACATGGTTGGCCCGCCTAGGTGGCTCCTCATCCTCCTCACAGTCATAGTCATCCAGGATGGGAGTCTCTGCCGCTGGGTCCCCAGGCGCGGGAGAGCTTCCGTTCATCCCGGCTGCAAACCCACAATTCCGTCATTGGCGTGCCCATCCGGGAGACTCCCATCCT\r\n",
      "+\r\n",
      "DFFFHHHHHJJJHHJJJJJJJJJJJJJIIJJJJJJHIGHFFFDDDDDDD@CDDDDDDDDDDDDDDDDDDDDDCCEEDEEEEDDDDDDDDDDDD?CDDBDF>FHHHHJHJJJJJJJJJJJJJGIEHJJHHHHFFFFFFEDDDDD=@CDDDDDDDDDDDDDDDDDDEDDDDDD<BD@CCB<BB<@9<CCDD<@CD3\r\n",
      "@SRR3503015.6 6 length=194\r\n",
      "TCCAAAGGAAGTGAGGGGTCCTTCTATGGTGAAACTAATTTGTGATTCATGTGTTCCCTCATCTGAGAATCTGGGGCAGGCCCATTTCCATGACCCTGGCTGACCTGCGCAACGTGATCAAGACACTGTTGAAAGCTGGCCACATAGCATGGACTTCTTCTGGAAAAGGGTCATGGAAATGGGCCTGCCCCAGA\r\n",
      "+\r\n",
      "DFFFHHHGHJJHIJJJJJIJJIJIIIJIIFHIJIJJJJJJJIIIHIJJJJJJJJJJJJJJJJJJJJIJJHHCHFHFFDDDDDDBDDDDEEEDCCDDDADDFHHHHHJJJJJIJJ:CHIIJJJJJJJJJJJGIIJIJJJJJJJJIHJIJJJJHHHHHHFFFFFFDEEEEDD?CDDDDCDDDDDDDDDDDDDDBD8\r\n",
      "@SRR3503015.7 7 length=194\r\n",
      "CCTTTTAAACCAGAGTCGAACTTTACTGTCTCCTCAGCAACAACAGCAGCAGCAAGTGGCACTTGGCCCTGGCATGCCAGCAAAGCCTCTTCAACACGGCTGGGTCTACGGTGTTTTGTTCCTTGCCCGTCAGGAGGAGGGTTGGACCCAGGGCTCCAGGGCTAGAAAAGTGTTGAAGAGGCTTTGCTGGCATG\r\n",
      "+\r\n",
      "DDFFHHGHGJJJJJJHIJJJJJJJIIJJGJIJJGIHIJIJIJJIGGGIJJJJIIJI@GHIIJIHHHHHFFDBCEEEEDDDDDDDCCDDDDDDEDDDBADFFHHHFHIIJJJFGHIJJJIJIJJJIJJJIIJIIJFHIFHHI;E@ECEHFFFDDCEDDDDDDDBDDCDCDD>AACBACDCDB@?CDCCCAC@CC<\r\n",
      "@SRR3503015.8 8 length=194\r\n",
      "GGTGGGGACAGGGGCGATGCTTCAGGTGGTGGGGATAGAGGCGTCTCAAGTGCAGGAGATGGGGGTGACTCTTCCGGTGGAGACAAGGGCGACTCCTCACCTGAGGAGTCACCCCTGTCCCCACCACCTGAGGAATCACCCACGTCCCCACCACCTGAGGCATCACGCCTGTCACCACCACCTGAGGAATTGCC\r\n",
      "+\r\n",
      "DDBDHHHHHJJJJJIJJJJJJJJIJJ8BH@GIHFEDEECEEDDBDDDEDDEDDDDDDDBDDDDDD9BDDDDCCDDDDDDDDDDDDDDDDDDDDDDDCDDFFHHHHHJJJJJJJJJJJJJJJJJJJJJJJJJJJJIJIJJJJJJHIJJJJJHHHFFFFEEDDBDDDDDDDDDDDDDDDDDDDDDDACDDDDDDDD\r\n",
      "@SRR3503015.9 9 length=194\r\n",
      "CAGTACATGCACTTCAGCACCTTCTGGGCGTCTTCCTTCCCTTCCATTTCCAGCAAGGAGCGTTTGCGAGGCTTGGACCAGCGCTTGGGGTTGTTGTCGCTGTAAGGACTGCAGCGCTGCCTACGACACCCTGGTGGAGTTGACGGTGCACATGAACGAGACGGGGCATTACCGCGACGACAACCATGAGACCG\r\n",
      "+\r\n",
      "DDFFHHHHHJIJJJFIJJJJJJJGIJJIGHHGHIGGIIJJIJJGHFHHIHGIIJJJJJEHIGEFFFFDBDDBBCB?CDDDDDDBDDBDDD59@>BDADDFFHHGHHJJJIJIJJJJIIJIJJIJJIJJJJJJJJCGIGIGHHHCDFBBDEEDDDCDDDDDDDDDDDBBCCDDDBDD9BDBDBDDDDCCCACC@>\r\n",
      "@SRR3503015.10 10 length=194\r\n",
      "ACACTGACCCATATGCTCAGCCCCCATTGACTCCTCGGCCCCAACCTCCGCCCCCTGAGAGCTGCTGTGCTCTGCCCCCTCGCTCACTGCCCTCCGCTAAAGGCAACTTCAGGGGGCTGGGGTCGGGGTGGCTTATGCAATGGGGCAAATGGGTCACGGGACTGAGGGCGTGAGGGTGGGCGAGAATAAGGGTC\r\n",
      "+\r\n",
      "ADEFHFHHGJJJJJJIJJIIJIJJJHHIGHJEHGDHGIIHIJCHJIIGIAGIGFFAB>ACDB;>AC@@ACA:@:ACBDD50&9@@7CC>CCD@BD9&FFFFHHHHHJJIJGIJJJJJIJJJJDGIJJI;=E;BFDEEEDECED?BB@DDDDDD>CBD>BD;>ABCCBDDD>B?BDD2?BD?B9@99@CCCCD<>\r\n"
     ]
    }
   ],
   "source": [
    "! cat srr3503015_sample.fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 3125152\r\n",
      "-rw-rw-r-- 1 tom tom 138245449 Mar 21  2009 chr10.fa\r\n",
      "-rw-rw-r-- 1 tom tom 137706654 Mar 21  2009 chr11.fa\r\n",
      "-rw-rw-r-- 1 tom tom     40929 Mar 21  2009 chr11_gl000202_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom 136528940 Mar 21  2009 chr12.fa\r\n",
      "-rw-rw-r-- 1 tom tom 117473283 Mar 21  2009 chr13.fa\r\n",
      "-rw-rw-r-- 1 tom tom 109496538 Mar 21  2009 chr14.fa\r\n",
      "-rw-rw-r-- 1 tom tom 104582027 Mar 21  2009 chr15.fa\r\n",
      "-rw-rw-r-- 1 tom tom  92161856 Mar 21  2009 chr16.fa\r\n",
      "-rw-rw-r-- 1 tom tom   1714462 Mar 21  2009 chr17_ctg5_hap1.fa\r\n",
      "-rw-rw-r-- 1 tom tom  82819122 Mar 21  2009 chr17.fa\r\n",
      "-rw-rw-r-- 1 tom tom     38271 Mar 21  2009 chr17_gl000203_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom     82960 Mar 21  2009 chr17_gl000204_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom    178103 Mar 21  2009 chr17_gl000205_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom     41845 Mar 21  2009 chr17_gl000206_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom  79638800 Mar 21  2009 chr18.fa\r\n",
      "-rw-rw-r-- 1 tom tom      4371 Mar 21  2009 chr18_gl000207_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom  60311570 Mar 21  2009 chr19.fa\r\n",
      "-rw-rw-r-- 1 tom tom     94566 Mar 21  2009 chr19_gl000208_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom    162376 Mar 21  2009 chr19_gl000209_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom 254235640 Mar 20  2009 chr1.fa\r\n",
      "-rw-rw-r-- 1 tom tom    108584 Mar 21  2009 chr1_gl000191_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom    558468 Mar 21  2009 chr1_gl000192_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom  64286038 Mar 21  2009 chr20.fa\r\n",
      "-rw-rw-r-- 1 tom tom  49092500 Mar 21  2009 chr21.fa\r\n",
      "-rw-rw-r-- 1 tom tom     28259 Mar 21  2009 chr21_gl000210_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom  52330665 Mar 21  2009 chr22.fa\r\n",
      "-rw-rw-r-- 1 tom tom 248063367 Mar 20  2009 chr2.fa\r\n",
      "-rw-rw-r-- 1 tom tom 201982885 Mar 20  2009 chr3.fa\r\n",
      "-rw-rw-r-- 1 tom tom    602251 Mar 21  2009 chr4_ctg9_hap1.fa\r\n",
      "-rw-rw-r-- 1 tom tom 194977368 Mar 20  2009 chr4.fa\r\n",
      "-rw-rw-r-- 1 tom tom    193607 Mar 21  2009 chr4_gl000193_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom    195321 Mar 21  2009 chr4_gl000194_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom 184533572 Mar 20  2009 chr5.fa\r\n",
      "-rw-rw-r-- 1 tom tom   4714751 Mar 21  2009 chr6_apd_hap1.fa\r\n",
      "-rw-rw-r-- 1 tom tom   4891294 Mar 21  2009 chr6_cox_hap2.fa\r\n",
      "-rw-rw-r-- 1 tom tom   4702619 Mar 21  2009 chr6_dbb_hap3.fa\r\n",
      "-rw-rw-r-- 1 tom tom 174537375 Mar 20  2009 chr6.fa\r\n",
      "-rw-rw-r-- 1 tom tom   4776945 Mar 21  2009 chr6_mann_hap4.fa\r\n",
      "-rw-rw-r-- 1 tom tom   4930081 Mar 21  2009 chr6_mcf_hap5.fa\r\n",
      "-rw-rw-r-- 1 tom tom   4704239 Mar 21  2009 chr6_qbl_hap6.fa\r\n",
      "-rw-rw-r-- 1 tom tom   5027155 Mar 21  2009 chr6_ssto_hap7.fa\r\n",
      "-rw-rw-r-- 1 tom tom 162321443 Mar 20  2009 chr7.fa\r\n",
      "-rw-rw-r-- 1 tom tom    186576 Mar 21  2009 chr7_gl000195_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom 149291309 Mar 21  2009 chr8.fa\r\n",
      "-rw-rw-r-- 1 tom tom     39715 Mar 21  2009 chr8_gl000196_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom     37941 Mar 21  2009 chr8_gl000197_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom 144037706 Mar 21  2009 chr9.fa\r\n",
      "-rw-rw-r-- 1 tom tom     91909 Mar 21  2009 chr9_gl000198_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom    173294 Mar 21  2009 chr9_gl000199_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom    190798 Mar 21  2009 chr9_gl000200_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom     36893 Mar 21  2009 chr9_gl000201_random.fa\r\n",
      "-rw-rw-r-- 1 tom tom     16909 Mar 21  2009 chrM.fa\r\n",
      "-rw-rw-r-- 1 tom tom    169914 Mar 21  2009 chrUn_gl000211.fa\r\n",
      "-rw-rw-r-- 1 tom tom    190612 Mar 21  2009 chrUn_gl000212.fa\r\n",
      "-rw-rw-r-- 1 tom tom    167540 Mar 21  2009 chrUn_gl000213.fa\r\n",
      "-rw-rw-r-- 1 tom tom    140489 Mar 21  2009 chrUn_gl000214.fa\r\n",
      "-rw-rw-r-- 1 tom tom    176012 Mar 21  2009 chrUn_gl000215.fa\r\n",
      "-rw-rw-r-- 1 tom tom    175756 Mar 21  2009 chrUn_gl000216.fa\r\n",
      "-rw-rw-r-- 1 tom tom    175608 Mar 21  2009 chrUn_gl000217.fa\r\n",
      "-rw-rw-r-- 1 tom tom    164386 Mar 21  2009 chrUn_gl000218.fa\r\n",
      "-rw-rw-r-- 1 tom tom    182798 Mar 21  2009 chrUn_gl000219.fa\r\n",
      "-rw-rw-r-- 1 tom tom    165055 Mar 21  2009 chrUn_gl000220.fa\r\n",
      "-rw-rw-r-- 1 tom tom    158521 Mar 21  2009 chrUn_gl000221.fa\r\n",
      "-rw-rw-r-- 1 tom tom    190615 Mar 21  2009 chrUn_gl000222.fa\r\n",
      "-rw-rw-r-- 1 tom tom    184081 Mar 21  2009 chrUn_gl000223.fa\r\n",
      "-rw-rw-r-- 1 tom tom    183303 Mar 21  2009 chrUn_gl000224.fa\r\n",
      "-rw-rw-r-- 1 tom tom    215413 Mar 21  2009 chrUn_gl000225.fa\r\n",
      "-rw-rw-r-- 1 tom tom     15325 Mar 21  2009 chrUn_gl000226.fa\r\n",
      "-rw-rw-r-- 1 tom tom    130958 Mar 21  2009 chrUn_gl000227.fa\r\n",
      "-rw-rw-r-- 1 tom tom    131719 Mar 21  2009 chrUn_gl000228.fa\r\n",
      "-rw-rw-r-- 1 tom tom     20328 Mar 21  2009 chrUn_gl000229.fa\r\n",
      "-rw-rw-r-- 1 tom tom     44581 Mar 21  2009 chrUn_gl000230.fa\r\n",
      "-rw-rw-r-- 1 tom tom     27950 Mar 21  2009 chrUn_gl000231.fa\r\n",
      "-rw-rw-r-- 1 tom tom     41482 Mar 21  2009 chrUn_gl000232.fa\r\n",
      "-rw-rw-r-- 1 tom tom     46876 Mar 21  2009 chrUn_gl000233.fa\r\n",
      "-rw-rw-r-- 1 tom tom     41358 Mar 21  2009 chrUn_gl000234.fa\r\n",
      "-rw-rw-r-- 1 tom tom     35180 Mar 21  2009 chrUn_gl000235.fa\r\n",
      "-rw-rw-r-- 1 tom tom     42789 Mar 21  2009 chrUn_gl000236.fa\r\n",
      "-rw-rw-r-- 1 tom tom     46801 Mar 21  2009 chrUn_gl000237.fa\r\n",
      "-rw-rw-r-- 1 tom tom     40754 Mar 21  2009 chrUn_gl000238.fa\r\n",
      "-rw-rw-r-- 1 tom tom     34517 Mar 21  2009 chrUn_gl000239.fa\r\n",
      "-rw-rw-r-- 1 tom tom     42788 Mar 21  2009 chrUn_gl000240.fa\r\n",
      "-rw-rw-r-- 1 tom tom     43012 Mar 21  2009 chrUn_gl000241.fa\r\n",
      "-rw-rw-r-- 1 tom tom     44410 Mar 21  2009 chrUn_gl000242.fa\r\n",
      "-rw-rw-r-- 1 tom tom     44224 Mar 21  2009 chrUn_gl000243.fa\r\n",
      "-rw-rw-r-- 1 tom tom     40744 Mar 21  2009 chrUn_gl000244.fa\r\n",
      "-rw-rw-r-- 1 tom tom     37401 Mar 21  2009 chrUn_gl000245.fa\r\n",
      "-rw-rw-r-- 1 tom tom     38934 Mar 21  2009 chrUn_gl000246.fa\r\n",
      "-rw-rw-r-- 1 tom tom     37167 Mar 21  2009 chrUn_gl000247.fa\r\n",
      "-rw-rw-r-- 1 tom tom     40598 Mar 21  2009 chrUn_gl000248.fa\r\n",
      "-rw-rw-r-- 1 tom tom     39289 Mar 21  2009 chrUn_gl000249.fa\r\n",
      "-rw-rw-r-- 1 tom tom 158375978 Mar 21  2009 chrX.fa\r\n",
      "-rw-rw-r-- 1 tom tom  60561044 Mar 21  2009 chrY.fa\r\n"
     ]
    }
   ],
   "source": [
    "!ls /home/tom/zdata/bioref/chromFa-hg19 -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Slideshow",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
