{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# data source\n",
    "- [illumina提供的打包好的参考序列、注释信息、index文件](http://blog.biochen.com/archives/279), [Source](http://support.illumina.com/sequencing/sequencing_software/igenome.html)\n",
    "\n",
    "- [hg19 from UCSC](http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/)\n",
    "\n",
    "\n",
    "# reference\n",
    "\n",
    "* [Integrated digital error suppression for improved detection of circulating tumor DNA](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4907374/), [Data](https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP069806)\n",
    "* [An ultrasensitive method for quantitating circulating tumor DNA with broad patient coverage](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4016134/)\n",
    "\n",
    "\n",
    "# tools\n",
    "\n",
    "* [awesome-bioinformation](https://github.com/danielecook/Awesome-Bioinformatics)\n",
    "\n",
    "### Genome Browser\n",
    "\n",
    "* [COSMIC Genome Browser](http://cancer-beta.sanger.ac.uk/cosmic/browse/genome) This site contains all the cancer genetic data curated into COSMIC, providing a genomic visualisation. Additional genomic annotations are provided to give a much broader perspective on the impact of these somatic mutations.\n",
    "\n",
    "* [JBrowser](http://gmod.org/wiki/JBrowse)JBrowse is a genome browser with a fully dynamic AJAX interface, being developed as the eventual successor to GBrowse. It is very fast and scales well to large datasets. JBrowse is javascript-based and does almost all of its work directly in the user's web browser, with minimal requirements for the server.\n",
    "    Features\n",
    "    - Fast, smooth scrolling and zooming. Explore your genome with unparalleled speed.\n",
    "    - Scales easily to multi-gigabase genomes and deep-coverage sequencing.\n",
    "    - Supports GFF3, BED, FASTA, Wiggle, BigWig, BAM, VCF (with tabix), REST, and more. BAM, BigWig, and VCF data are displayed directly from the compressed binary file with no conversion needed.\n",
    "    - Very light server resource requirements. In fact, JBrowse has no back-end server code, just tools for formatting data files to be read directly over HTTP. Serve huge datasets from a single low-cost cloud instance.\n",
    "    \n",
    "* [Plugins for JBrowser](https://gmod.github.io/jbrowse-registry/#!?page=1&perPage=5)\n",
    "\n",
    "## python\n",
    "* [awesome-python](https://github.com/vinta/awesome-python)\n",
    "* [bioconda](https://anaconda.org/bioconda) packages managed by conda, [Github](https://github.com/bioconda)\n",
    "* [bcbb](https://github.com/chapmanb/bcbb)Incubator for useful bioinformatics code, primarily in Python and R\n",
    "\n",
    "\n",
    "## C++\n",
    "\n",
    "* [awesome C++](https://github.com/fffaraz/awesome-cpp)A curated list of awesome C/C++ frameworks, libraries, resources, and shiny things. Inspired by awesome-... stuff.\n",
    "* [SeqAn](https://github.com/seqan/seqan), [SeqAn3](https://github.com/seqan/seqan3)The modern C++ library for sequence analysis. Contains version 3 of the library and API docs.\n",
    "\n",
    "\n",
    "## R\n",
    "* [R notebook](http://rmarkdown.rstudio.com/r_notebooks.html) \n",
    "\n",
    "\n",
    "## organization , lab, person\n",
    "\n",
    "* [The Lab for Data Intensive Biology @ Michigan State University](https://github.com/dib-lab)\n",
    "* [Bioinformatics Core at Harvard T.H. Chan School of Public Health](https://github.com/chapmanb), [BLOG](http://bcb.io/), Blue Collar Bioinformatics (bcbio) is community-developed variant calling, RNA-seq and small RNA analysis.\n",
    "* [Bioinformatics Team (BioITeam) at the University of Texas](https://wikis.utexas.edu/display/bioiteam/Variant+calling+using+SAMtools),The BioITeam (bio - i - team, short for bioinformatics team) is a consortium of bio-computing scientists at the University of Texas comprised of scientists from TACC, CCBB, CSSB, GSAF, TI-3D, and SSC.\n",
    "\n",
    "* [COSMIC](http://cancer-beta.sanger.ac.uk/cosmic) the Catalogue Of Somatic Mutations In Cancer, is the world's largest and most comprehensive resource for exploring the impact of somatic mutations in human cancer. [ref: COSMIC (the Catalogue of Somatic Mutations in Cancer): a resource to investigate acquired mutations in human cancer](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2808858/)\n",
    "\n",
    "\n",
    "* [Generic Model Organism Database project](http://gmod.org/wiki/Main_Page), Generic Model Organism Database project, a collection of open source software tools for managing, visualising, storing, and disseminating genetic and genomic data.\n",
    "\n",
    "* [National Cancer Institute at the National Institutes of Health](https://www.cancer.gov/)\n",
    "\n",
    "* [The National Cancer Institute’s (NCI’s) Genomic Data Commons (GDC)](https://gdc.cancer.gov)The National Cancer Institute’s (NCI’s) Genomic Data Commons (GDC) is a data sharing platform that promotes precision medicine in oncology. It is not just a database or a tool; it is an expandable knowledge network supporting the import and standardization of genomic and clinical data from cancer research programs.\n",
    "\n",
    "* [Analysis tool used by NCI](https://tcga-data.nci.nih.gov/docs/publications/tcga/analyticaltools.html)\n",
    "\n",
    "* [National Cancer Informatics Program](https://github.com/NCIP)\n",
    "\n",
    "\n",
    "# data file format\n",
    "* abif    : Applied Biosystem's sequencing trace format\n",
    "* ace     : Reads the contig sequences from an ACE assembly file.\n",
    "* embl    : The EMBL flat file format. Uses Bio.GenBank internally.\n",
    "* fasta   : The generic sequence file format where each record starts with an identifer line starting with a \">\" character, followed by lines of sequence.\n",
    "* fastq   : A \"FASTA like\" format used by Sanger which also stores PHRED sequence quality values (with an ASCII offset of 33).\n",
    "* fastq-sanger : An alias for \"fastq\" for consistency with BioPerl and EMBOSS\n",
    "* fastq-solexa : Original Solexa/Illumnia variant of the FASTQ format which encodes Solexa quality scores (not PHRED quality scores) with an ASCII offset of 64.\n",
    "* fastq-illumina - Solexa/Illumina 1.3 to 1.7 variant of the FASTQ format which encodes PHRED quality scores with an ASCII offset of 64 (not 33). Note as of version 1.8 of the CASAVA pipeline Illumina will produce FASTQ files using the standard Sanger encoding.\n",
    "* genbank : The GenBank or GenPept flat file format.\n",
    "* gb      : An alias for \"genbank\", for consistency with NCBI Entrez Utilities\n",
    "* ig      : The IntelliGenetics file format, apparently the same as the MASE alignment format.\n",
    "* imgt    : An EMBL like format from IMGT where the feature tables are more indented to allow for longer feature types.\n",
    "* pdb-seqres :  Reads a Protein Data Bank (PDB) file to determine the complete protein sequence as it appears in the header (no dependencies).\n",
    "* pdb-atom : Uses Bio.PDB to determine the (partial) protein sequence as it appears in the structure based on the atom coordinate section of the file (requires NumPy for Bio.PDB).\n",
    "* phd     : Output from PHRED, used by PHRAP and CONSED for input.\n",
    "* pir     : A \"FASTA like\" format introduced by the National Biomedical Research Foundation (NBRF) for the Protein Information Resource (PIR) database, now part of UniProt.\n",
    "* seqxml  : SeqXML, simple XML format described in Schmitt et al (2011).\n",
    "* sff     : Standard Flowgram Format (SFF), typical output from Roche 454.\n",
    "* sff-trim: Standard Flowgram Format (SFF) with given trimming applied.\n",
    "* swiss   : Plain text Swiss-Prot aka UniProt format.\n",
    "* tab     : Simple two column tab separated sequence files, where each line holds a record's identifier and sequence. For example, this is used as by Aligent's eArray software when saving microarray probes in a minimal tab delimited text file.\n",
    "* qual    : A \"FASTA like\" format holding PHRED quality values from sequencing DNA, but no actual sequences (usually provided in separate FASTA files).\n",
    "* uniprot-xml * The UniProt XML format (replacement for the SwissProt plain text format which we call \"swiss\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
